crypt_test.go

```diff
@@ -120,162 +118,208 @@ func (r *mockAuthRequestor) RequestRecoveryKey(volumeName, sourceDevicePath stri
 	}
 }
 
-type cryptSuite struct {
-	cryptTestBase
-	keyDataTestBase
-	testutil.KeyringTestBase
+// mockLUKS2Container represents a LUKS2 container and its associated state
+type mockLUKS2Container struct {
+	keyslots map[int][]byte
+}
 
-	mockKeyslotsDir   string
-	mockKeyslotsCount int
+func (c *mockLUKS2Container) nextFreeSlot() (slot int) {
+	var slots []int
+	for i := range c.keyslots {
+		slots = append(slots, i)
+	}
+	sort.Ints(slots)
+	for _, i := range slots {
+		if i != slot {
+			break
+		}
+		slot++
+	}
+	return slot
+}
 
-	mockLUKS2ActivateCalls []struct {
-		volumeName       string
-		sourceDevicePath string
+// mockLUKS2 mocks a device's global LUKS2 state. It provides mock
+// implementations of the various LUKS2 operations.
+type mockLUKS2 struct {
+	operations []string                       // A log of LUKS2 operations recorded during a test
```

Comment:
what we do sometimes is having a generic op structure to cover possible arg combinations and then have a slice of those:

```
ops []op

type op struct {
  op string
  paths []string
  slot int
  addKeyOptions *luks2.AddKeyOptions
 // etc ...
}
```

it doesn't matter too much, the main issue is that different structs might have the same Sprint output

Reply:
Thanks - I'll take a look at this at the end of this series of PRs